mlir/lib/Dialect/PDL/IR/Builtins.cpp

```diff
@@ -51,52 +52,172 @@ mlir::Attribute addElemToArrayAttr(mlir::PatternRewriter &rewriter,
   return rewriter.getArrayAttr(values);
 }
 
-LogicalResult add(mlir::PatternRewriter &rewriter, mlir::PDLResultList &results,
-                  llvm::ArrayRef<mlir::PDLValue> args) {
-  assert(args.size() == 2 && "Expected 2 arguments");
+template <UnaryOpKind T>
+LogicalResult unaryOp(PatternRewriter &rewriter, PDLResultList &results,
+                      ArrayRef<PDLValue> args) {
+  assert(args.size() == 1 && "Expected one operand for unary operation");
+  auto operandAttr = args[0].cast<Attribute>();
+
+  if (auto operandIntAttr = dyn_cast_or_null<IntegerAttr>(operandAttr)) {
+    auto integerType = cast<IntegerType>(operandIntAttr.getType());
+    auto bitWidth = integerType.getIntOrFloatBitWidth();
+
+    if constexpr (T == UnaryOpKind::exp2) {
+      uint64_t resultVal =
+          integerType.isUnsigned() || integerType.isSignless()
+              ? std::pow(2, operandIntAttr.getValue().getZExtValue())
+              : std::pow(2, operandIntAttr.getValue().getSExtValue());
+
+      APInt resultInt(bitWidth, resultVal, integerType.isSigned());
+
+      bool isOverflow = integerType.isSigned()
+                            ? resultInt.slt(operandIntAttr.getValue())
+                            : resultInt.ult(operandIntAttr.getValue());
+
+      if (isOverflow)
+        return failure();
+
+      results.push_back(rewriter.getIntegerAttr(integerType, resultInt));
+    } else if constexpr (T == UnaryOpKind::log2) {
+      auto getIntegerAsAttr = [&](const APSInt &value) {
+        int32_t log2Value = value.exactLogBase2();
+        assert(log2Value >= 0 &&
+               "log2 of an integer is expected to return an exact integer.");
+        return rewriter.getIntegerAttr(
+            integerType,
+            APSInt(APInt(bitWidth, log2Value), integerType.isUnsigned()));
+      };
+      // for log2 we treat signless integer as signed
+      if (integerType.isSignless())
+        results.push_back(
+            getIntegerAsAttr(APSInt(operandIntAttr.getValue(), false)));
+      else
+        results.push_back(getIntegerAsAttr(operandIntAttr.getAPSInt()));
+    } else {
+      llvm::llvm_unreachable_internal(
+          "encountered an unsupported unary operator");
+      return failure();
+    }
+    return success();
+  }
+
+  if (auto operandFloatAttr = dyn_cast_or_null<FloatAttr>(operandAttr)) {
+    auto floatType = operandFloatAttr.getType();
+
+    if constexpr (T == UnaryOpKind::exp2) {
+      auto maxVal = APFloat::getLargest(llvm::APFloat::IEEEhalf());
```

Comment:
Why do we use `IEEEhalf` here?

Reply:
Thanks for catching it. It should be handled case by case depending on float type.